### ISSUE TITLE
Segfaults/bus errors with mimalloc on M1 hardware in mimalloc > 1.7.2

### ISSUE THREAD

**Issue by jmather-sesi** [microsoft/mimalloc#557]
I am attempting to run the benchmarks provided in mimalloc-bench, however I am running into constant segfaults and bus errors when trying to use the default version of mimalloc (1.7.5).

The mimalloc `make test` passes. Unfortunately I have been unable to attach a debugger to any of the tests to inspect where it's crashing as it always seems to run without issues when under the debugger. The command that I'm using is `DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib lldb ./cfrac -- 17545186520507317056371138836327483792789528`

This seems to happen in mimalloc 1.7.3 and 1.7.5 - 1.7.2 does *not* have this issue. This is on macOS Monterey 12.2.1.

Running the printed execution commands directly in the terminal will result in either a segfault or a bus error.

The runs all look like the following (multiple runs snipped for brevity):
```
Users-MacBook-Pro:bench user$ ../../bench.sh -r=1 -n=2 mi allt
benchmarking on 10 cores.
use '-h' or '--help' for help on configuration options.

allocators:  mi
tests     :  cfrac espresso barnes larson-sized mstress rptest gs alloc-test xmalloc-test cscratch glibc-simple glibc-thread
(excluded tests:  sh6bench sh8bench redis sed lean lean-mathlib)

---- 1: cfrac

run 1: cfrac mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./cfrac 17545186520507317056371138836327483792789528
Command terminated by signal 11
cfrac       mi    0:00.02 116880 0.01 0.00 0 7486

---- 1: espresso

run 1: espresso mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./espresso ../../bench/espresso/largest.espresso
Command terminated by signal 11
espresso    mi    0:00.01 22832 0.00 0.00 0 1608

---- 1: barnes

run 1: barnes mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./barnes
warning: this program uses gets(), which is unsafe.
barnes      mi    0:01.56 78048 1.45 0.02 0 5058

---- 1: larsonN-sized

run 1: larsonN-sized mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./larson-sized 5 8 1000 5000 100 4141 10
larsonN-sized,mi, relative time: 3.108s
larsonN-sized mi    0:3.108 154464 49.19 0.30 0 16273

---- 1: mstressN

run 1: mstressN mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./mstress 10 50 25
Command terminated by signal 11
mstressN    mi    0:00.09 23616 0.00 0.00 0 1660

---- 1: rptestN

run 1: rptestN mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./rptest 10 0 1 2 500 1000 100 8 16000
rptestN,mi: ops/sec: 3719211, relative time: .537s
rptestN     mi    0:.537 114896 0.82 0.60 0 7408

---- 1: gs

run 1: gs mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib gs -dBATCH -dNODISPLAY /Users/user/dev/mimalloc-bench/extern/325462-sdm-vol-1-2abcd-3abcd.pdf
Command terminated by signal 11
gs          mi    0:00.56 31984 0.01 0.03 0 2698

---- 1: alloc-test1

run 1: alloc-test1 mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./alloc-test 1
threads: 1
alloc-test1 mi    0:03.07 31200 2.96 0.01 0 2130

---- 1: alloc-testN

run 1: alloc-testN mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./alloc-test 10
threads: 10
alloc-testN mi    0:03.33 33616 32.10 0.05 0 2290

---- 1: xmalloc-testN

run 1: xmalloc-testN mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./xmalloc-test -w 10 -t 5 -s 64
xmalloc-testN,mi, relative time: s
Command terminated by signal 11
xmalloc-testN mi    0: 25232 0.00 0.00 0 1766

---- 1: cache-scratch1

run 1: cache-scratch1 mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./cache-scratch 1 1000 1 2000000 10
cache-scratch1 mi    0:00.99 21920 0.86 0.01 0 1550

---- 1: cache-scratchN

run 1: cache-scratchN mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./cache-scratch 10 1000 1 2000000 10
cache-scratchN mi    0:00.13 22080 1.04 0.00 0 1569

---- 1: glibc-simple

run 1: glibc-simple mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./glibc-simple
Command terminated by signal 11
glibc-simple mi    0:00.29 43728 0.01 0.01 0 2913

---- 1: glibc-thread

run 1: glibc-thread mi: DYLD_INSERT_LIBRARIES=/Users/user/dev/mimalloc-bench/extern/mimalloc/out/release/libmimalloc.dylib ./glibc-thread 10
(standard_in) 1: parse error
glibc-thread,mi: iterations: , relative time: s
Command terminated by signal 11
glibc-thread mi    0: 28352 0.01 0.02 0 1987

results written to: /Users/user/dev/mimalloc-bench/out/bench/benchres.csv

#------------------------------------------------------------------
# test    alloc   time  rss    user  sys  page-faults page-reclaims

Command terminated by signal 11
cfrac       mi    00.02 116880 0.01 0.00 0 7486

Command terminated by signal 11
espresso    mi    00.01 22832 0.00 0.00 0 1608

barnes      mi    01.56 78048 1.45 0.02 0 5058

larsonN-sized mi    3.108 154464 49.19 0.30 0 16273

Command terminated by signal 11
mstressN    mi    00.09 23616 0.00 0.00 0 1660

rptestN     mi    .537 114896 0.82 0.60 0 7408

Command terminated by signal 11
gs          mi    00.56 31984 0.01 0.03 0 2698

alloc-test1 mi    03.07 31200 2.96 0.01 0 2130

alloc-testN mi    03.33 33616 32.10 0.05 0 2290

Command terminated by signal 11
xmalloc-testN mi     25232 0.00 0.00 0 1766

cache-scratch1 mi    00.99 21920 0.86 0.01 0 1550

cache-scratchN mi    00.13 22080 1.04 0.00 0 1569

Command terminated by signal 11
glibc-simple mi    00.29 43728 0.01 0.01 0 2913

Command terminated by signal 11
glibc-thread mi     28352 0.01 0.02 0 1987
```

**Comment by daanx** [comment]
Ah that is not good -- I cannot reproduce the crash (Mac air M1, Monterey 12.2.1) but was able to trigger an assertion failure in full debug build. Can you try to change line `alloc-override.c:84` (in the `dev` branch (~v1.7.5)):
```C
 #ifdef MI_OSX_ZONE
```
to 
```C
 #ifdef XMI_OSX_ZONE
```
and see if that fixes the issue for you? (it does for me). I would still want to find the root cause (perhaps things changed in Monterey?) but if this works that is both a good fix and gives a good indication what went wrong.

**Comment by jmather-sesi** [comment]
When I run cfrac with the debug library, I get the following:

```mimalloc: warning: mi_free: pointer might not point to a valid heap region: 0x6000033cc020
(this may still be a valid very large allocation (over 64MiB))
mimalloc: error: mi_free: pointer does not point to a valid heap space: 0x6000033cc020
mimalloc: assertion failed: at "/Users/prisms/mimalloc-bench/extern/mimalloc/include/mimalloc-internal.h":427, _mi_segment_page_idx_of
  assertion: "idx < segment->capacity"
Abort trap: 6
```

Changing MI_OSX_ZONE to XMI_OSX_ZONE seems to have resolved the crashes.

**Comment by DavidWzh** [comment]
heap stats:    peak      total      freed    current       unit      count   
normal   1:    7.3 KiB  496.8 KiB  489.6 KiB    7.1 KiB      8 B     63.3 K    not all freed!
normal   4:   99.3 KiB    3.9 MiB    3.8 MiB   97.7 KiB     32 B    129.1 K    not all freed!
normal   6:   50.5 KiB   15.2 MiB   15.1 MiB   31.6 KiB     48 B    332.3 K    not all freed!
normal   8:   39.5 KiB    2.7 MiB    2.7 MiB   26.6 KiB     64 B     45.6 K    not all freed!
normal   9:   20.8 KiB    5.2 MiB    5.2 MiB   15.2 KiB     80 B     68.9 K    not all freed!
normal  10:   11.1 KiB    1.8 MiB    1.8 MiB    9.0 KiB     96 B     19.9 K    not all freed!
normal  11:   22.6 KiB   16.2 MiB   16.2 MiB   21.3 KiB    112 B    152.1 K    not all freed!
normal  12:   32.0 KiB    1.9 MiB    1.8 MiB   22.4 KiB    128 B     15.5 K    not all freed!
normal  13:   26.6 KiB   11.0 MiB   11.0 MiB   15.2 KiB    160 B     72.4 K    not all freed!
normal  14:   43.2 KiB    1.9 MiB    1.8 MiB   42.7 KiB    192 B     10.5 K    not all freed!
normal  15:  290.3 KiB    1.8 MiB    1.5 MiB  289.6 KiB    224 B      8.6 K    not all freed!
normal  16:    9.7 KiB    1.6 MiB    1.6 MiB    9.2 KiB    256 B      6.8 K    not all freed!
normal  17:   75.9 KiB   18.2 MiB   18.1 MiB   74.9 KiB    320 B     59.8 K    not all freed!
normal  18:   41.7 KiB    1.9 MiB    1.9 MiB   38.4 KiB    384 B      5.3 K    not all freed!
normal  19:   18.0 KiB    1.2 MiB    1.2 MiB   17.1 KiB    448 B      3.0 K    not all freed!
normal  20:    5.0 KiB    1.3 MiB    1.3 MiB    4.0 KiB    512 B      2.6 K    not all freed!
normal  21:   52.7 KiB   20.2 MiB   20.2 MiB    8.1 KiB    640 B     33.2 K    not all freed!
normal  22:   53.4 KiB    2.1 MiB    2.0 MiB   51.2 KiB    768 B      2.9 K    not all freed!
normal  23:  120.3 KiB    3.0 MiB    2.9 MiB   79.0 KiB    896 B      3.5 K    not all freed!
normal  24:   55.2 KiB    1.6 MiB    1.6 MiB   53.2 KiB    1.0 KiB    1.7 K    not all freed!
normal  25:  215.8 KiB   18.6 MiB   18.5 MiB  125.4 KiB    1.2 KiB   15.2 K    not all freed!
normal  26:   24.0 KiB    2.7 MiB    2.7 MiB   21.0 KiB    1.5 KiB    1.8 K    not all freed!
normal  27:   28.1 KiB    2.8 MiB    2.7 MiB   24.5 KiB    1.7 KiB    1.6 K    not all freed!
normal  28:   32.1 KiB    2.8 MiB    2.7 MiB   28.1 KiB    2.0 KiB    1.4 K    not all freed!
normal  29:  102.9 KiB  312.9 MiB  312.8 MiB   92.8 KiB    2.5 KiB  128.1 K    not all freed!
normal  30:  231.9 KiB    6.6 MiB    6.4 MiB  225.8 KiB    3.0 KiB    2.2 K    not all freed!
normal  31:  112.4 KiB    6.0 MiB    5.9 MiB  105.4 KiB    3.5 KiB    1.7 K    not all freed!
normal  32:   46.6 MiB  637.2 MiB  590.7 MiB   46.5 MiB    4.0 KiB  163.1 K    not all freed!
normal  33:  331.2 KiB   17.1 MiB   16.8 MiB  306.1 KiB    5.0 KiB    3.5 K    not all freed!
normal  34:  228.8 KiB   15.1 MiB   15.0 MiB  162.6 KiB    6.0 KiB    2.5 K    not all freed!
normal  35:  196.7 KiB   14.3 MiB   14.1 MiB  196.7 KiB    7.0 KiB    2.0 K    not all freed!
normal  36:  843.2 KiB   16.3 MiB   15.5 MiB  843.2 KiB    8.0 KiB    2.0 K    not all freed!
normal  37:    2.7 MiB   57.2 MiB   56.8 MiB  401.5 KiB   10.0 KiB    5.8 K    not all freed!
normal  38:  301.1 KiB   38.5 MiB   38.2 MiB  289.1 KiB   12.0 KiB    3.2 K    not all freed!
normal  39:  491.9 KiB   43.9 MiB   43.4 MiB  477.8 KiB   14.0 KiB    3.2 K    not all freed!
normal  40:    1.5 MiB   46.3 MiB   45.7 MiB  530.0 KiB   16.0 KiB    2.9 K    not all freed!
normal  41:  642.5 KiB  100.6 MiB  100.0 MiB  622.4 KiB   20.0 KiB    5.1 K    not all freed!
normal  42:  650.5 KiB   98.5 MiB   97.8 MiB  626.4 KiB   24.0 KiB    4.2 K    not all freed!
normal  43:  758.9 KiB  114.7 MiB  113.9 MiB  730.8 KiB   28.1 KiB    4.1 K    not all freed!
normal  44:  835.2 KiB  109.8 MiB  109.1 MiB  738.8 KiB   32.1 KiB    3.5 K    not all freed!
normal  45:    3.7 MiB  230.0 MiB  226.3 MiB    3.7 MiB   40.1 KiB    5.8 K    not all freed!
normal  46:  722.8 KiB  130.9 MiB  130.3 MiB  626.4 KiB   48.1 KiB    2.7 K    not all freed!
normal  47:    1.1 MiB   59.3 MiB   58.2 MiB    1.0 MiB   56.2 KiB    1.0 K    not all freed!
normal  48:    1.1 MiB   11.8 MiB   10.6 MiB    1.1 MiB   64.2 KiB    190      not all freed!
normal  49:    1.6 MiB    9.2 MiB    7.6 MiB    1.5 MiB   80.3 KiB    118      not all freed!
normal  50:    1.5 MiB    3.9 MiB    2.4 MiB    1.5 MiB   96.3 KiB     42      not all freed!
normal  51:    1.7 MiB    4.1 MiB    2.5 MiB    1.6 MiB  112.4 KiB     38      not all freed!
normal  52:    1.1 MiB    3.0 MiB    1.8 MiB    1.1 MiB  128.5 KiB     24      not all freed!
normal  53:    1.7 MiB    9.3 MiB    7.6 MiB    1.7 MiB  160.6 KiB     60      not all freed!
normal  54:    1.3 MiB    3.5 MiB    2.4 MiB    1.1 MiB  192.7 KiB     19      not all freed!
normal  55:    2.1 MiB    4.8 MiB    2.8 MiB    1.9 MiB  224.8 KiB     22      not all freed!
normal  56:    2.2 MiB    9.7 MiB    7.5 MiB    2.2 MiB  257.0 KiB     39      not all freed!
normal  57:    2.1 MiB    5.9 MiB    3.7 MiB    2.1 MiB  321.2 KiB     19      not all freed!
normal  58:    1.5 MiB    2.6 MiB    1.5 MiB    1.1 MiB  385.5 KiB      7      not all freed!
normal  59:    2.6 MiB    6.1 MiB    3.5 MiB    2.6 MiB  449.7 KiB     14      not all freed!
normal  60:    2.5 MiB    4.5 MiB    2.0 MiB    2.5 MiB  514.0 KiB      9      not all freed!
normal  61:    4.3 MiB    9.3 MiB    5.0 MiB    4.3 MiB  642.5 KiB     15      not all freed!
normal  62:    2.2 MiB    4.5 MiB    2.2 MiB    2.2 MiB  771.0 KiB      6      not all freed!
normal  63:    7.0 MiB   11.3 MiB    4.3 MiB    7.0 MiB  899.5 KiB     13      not all freed!
normal  64:    3.0 MiB    4.0 MiB    1.0 MiB    3.0 MiB    1.0 MiB      4      not all freed!
normal  65:    2.5 MiB    5.0 MiB    2.5 MiB    2.5 MiB    1.2 MiB      4      not all freed!
normal  67:    3.5 MiB    3.5 MiB      0        3.5 MiB    1.7 MiB      2      not all freed!

heap stats:    peak      total      freed    current       unit      count   
    normal:  105.0 Mi     2.2 Gi     2.1 Gi   104.2 Mi     1.6 KiB    1.4 M    not all freed!
      huge:   17.7 Mi    23.7 Mi     6.0 Mi    17.7 Mi     5.9 MiB      4      not all freed!
     giant:      0          0          0          0                            ok
     total:  122.8 MiB    2.2 GiB    2.1 GiB  122.0 MiB                        not all freed!
malloc req:  117.0 MiB    2.1 GiB    2.0 GiB  116.2 MiB                        not all freed!

  reserved:  256.0 MiB  256.0 MiB      0      256.0 MiB                        not all freed!
 committed:  256.0 MiB  256.0 MiB      0      256.0 MiB                        not all freed!
     reset:   12.0 MiB   23.4 MiB   26.7 MiB   -3.2 MiB                        ok
   touched:  147.3 MiB  270.6 MiB  123.9 MiB  146.7 MiB                        not all freed!
  segments:     36         63         27         36                            not all freed!
-abandoned:      0          0          0          0                            ok
   -cached:      0          0          0          0                            ok
     pages:    974        1.7 Ki     801        962                            not all freed!
-abandoned:      0          0          0          0                            ok
 -extended:    1.9 Ki 
 -noretire:    1.8 Ki 
     mmaps:      1    
   commits:      0    
   threads:      0          0          0          0                            ok
  searches:     1.6 avg
numa nodes:       1
   elapsed:     597.917 s
   process: user: 11.644 s, system: 0.616 s, faults: 2, rss: 151.6 MiB, commit: 256.0 MiB
mimalloc: process done: 0xffff88170e20

**Comment by jmather-sesi** [comment]
I don't think the message above has anything to do with this issue. See: https://github.com/microsoft/mimalloc/issues/565

**Comment by daanx** [comment]
@jmather-sesi : apologies for not getting back earlier -- it is clear what the issue is and I will fix this soon in the new release of mimalloc.  (and indeed, this is not related at all to #565).

**Comment by daanx** [comment]
I think I fixed the root cause of these errors. Can you try again the latest dev-slice (or dev) branch (and compile as C (not C++) as is done by default).

**Comment by jmather-sesi** [comment]
Apologies for not getting back to you on this. This seems like it has been resolved in 1.7.6. Thanks for your efforts!
